### PR TITLE
[embind] Fix cross-thread check for deleted vals

### PR DESCRIPTION
**Pull request by RReverser**

### CODE REVIEW

**Comment by RReverser** [comment]
@sbc100 @brendandahl Could you take a quick look please? I missed this edge case in #20344 and would rather it doesn't get into release unfixed 😬

File: test/test_core.py [hunk 1]
```diff
@@ -7804,6 +7803,34 @@ def test_embind_val_cross_thread(self):
...
+          val& value = opt_value.emplace(1);
+          // Move out from the optional storage so that we free the value on the same thread.
+          val moved_out = std::move(value);
+        }).join();
+        // Now std::optional is initialized but with a deleted value handle.
+        // There should be no cross-thread error here when it tries to free that value,
+        // because the value has already been deleted on the correct thread.
+      }
+    ''')
+    self.do_runf('test_embind_val_cross_thread.cpp')
```
Comment (sbc100) [inline test/test_core.py:7832]: I wonder if both of these tests should be in test_other instead.. then they don't get run in so many configurations and use less CI time?

**Review by sbc100 — APPROVED** [review]

File: test/test_core.py (hunk 1 above)
Comment (RReverser) [reply test/test_core.py:7832]: Hm most Embind tests live in core, I assumed it's intentional? But I don't know the actual rules and usually just follow the precedent.

Comment (RReverser) [reply test/test_core.py:7832]: TIL. In that case, yeah, quite a lot of Embind tests could live in other.

Comment (sbc100) [reply test/test_core.py:7832]: OK with me to leave that as a followup.

**Comment by RReverser** [comment]
Those flaky tests... :( @sbc100 I already tried rerun, but just a different set of flaky tests started failing. Can you force-merge please?

**Pull request merged**